### PR TITLE
add `is_safe` to rust UUID

### PR DESCRIPTION
**Pull request by trim21**
I know little about rust, hope this is not a very bad idea.

close https://github.com/aminalaee/uuid-utils/issues/69

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -306,6 +307,18 @@ impl UUID {
             uuid: Uuid::from_u128(int),
         })
     }
+
+    #[getter]
+    fn is_safe<'py>(&self, py: Python<'py>) -> Bound<'py, PyAny> {
+        unsafe {
+            match SAFE_UUID {
```
Comment (aminalaee) [inline src/lib.rs:314]: Can't we implement the logic in Rust? Instead of using unsafe with GIL  which will slow us down?

Comment (trim21) [reply src/lib.rs]: I think we can return `*mut ffi::PyObject` directly, but I guess pyo3 will still bind it somewhere

**Comment by trim21** [comment]
also, are all of our generators safe?

**Comment by aminalaee** [comment]
I'm not sure if the whole approach is correct.
Looking at the Python and C source code, the `is_safe` is determined that when generating a UUID. I don't think just passing the UUID to the Python Enum will determine if it is safe or not.

**Comment by trim21** [comment]
> I'm not sure if the whole approach is correct. Looking at the Python and C source code, the `is_safe` is determined that when generating a UUID. I don't think just passing the UUID to the Python Enum will determine if it is safe or not.

what's why I ask https://github.com/aminalaee/uuid-utils/pull/70#issuecomment-2351729311 for many times, are our generators multiprocess safe?

uuid v3 and uuid v5 are hash based so thy are safe. uuid v1 and v6 are time based, and uuid v7 are time and random based, are they safe of not?

If they are all safe, we can just return `SafeUUIF.safe`. otherwise we need to pass then in uuid generator function.

**Comment by aminalaee** [comment]
I don't think it's that straightforward. Even our `uuid4` can be unsafe:

```py
import sys
import uuid_utils
sys.modules["uuid"] = uuid_utils
import os

import uuid

a = uuid.uuid4()
print(f"a = {a}")
os.fork()
b = uuid.uuid4()
print(f"b = {b}")
```

But what I think we can do is:
- Define SafeUUID Enum in Rust
- For now just return Unknown until a better approach is found

**Comment by trim21** [comment]
I find that stdlib uuid return unknown for event v3

File: python/uuid_utils/__init__.pyi [hunk 2]
```diff
@@ -188,3 +183,25 @@ RESERVED_NCS: str
...
+    "NAMESPACE_DNS",
+    "NAMESPACE_OID",
+    "NAMESPACE_URL",
+    "NAMESPACE_X500",
+    "RESERVED_FUTURE",
+    "RESERVED_MICROSOFT",
+    "RESERVED_NCS",
+    "RFC_4122",
+    "UUID",
+    "SafeUUID",
```
Comment (trim21) [inline python/uuid_utils/__init__.pyi:197]: need to add `uuid.SafeUUID` to make type checker consider it's exported from this module

**Comment by trim21** [comment]
`.into_ptr()` will wrap ptr with `ManuallyDrop` to avoid dec_ref on drop, keep refcnt correct.

**Comment by aminalaee** [comment]
I will do some benchmark tests locally and get back to this PR. Sorry it took so long.

**Comment by trim21** [comment]
it's a benchmark for `AtomicPtr.load`...

**Comment by aminalaee** [comment]
not really, it's a benchmark how PyO3 handles this and how it will affect our benchmarks.

**Comment by trim21** [comment]
it will add a new `PyMethodDefType::Getter` to method defs

```diff
9c9
<     prelude::*, pyclass::CompareOp, types::{PyBytes, PyDict},
---
>     ffi, prelude::*, pyclass::CompareOp, types::{PyBytes, PyDict},
12,13d11
< use std::hash::Hasher;
< use std::sync::atomic::{AtomicU64, Ordering};
14a13,14
> use std::{hash::Hasher, sync::atomic::AtomicPtr};
> use std::{ptr::null_mut, sync::atomic::{AtomicU64, Ordering}};
412a413,415
>     fn is_safe(&self) -> *mut ffi::PyObject {
>         return SAFE_UUID_UNKNOWN.load(Ordering::Relaxed);
>     }
816a820,828
>                 ::pyo3::impl_::pyclass::MaybeRuntimePyMethodDef::Static(
>                     ::pyo3::class::PyMethodDefType::Getter(
>                         ::pyo3::class::PyGetterDef::new(
>                             c"is_safe",
>                             UUID::__pymethod_get_is_safe__,
>                             c"",
>                         ),
>                     ),
>                 ),
1923a1936,1954
>     unsafe fn __pymethod_get_is_safe__(
>         py: ::pyo3::Python<'_>,
>         _slf: *mut ::pyo3::ffi::PyObject,
>     ) -> ::pyo3::PyResult<*mut ::pyo3::ffi::PyObject> {
>         #[allow(clippy::let_unit_value)]
>         let mut holder_0 = ::pyo3::impl_::extract_argument::FunctionArgumentHolder::INIT;
>         let result = ::pyo3::callback::convert(
>             py,
>             UUID::is_safe(
>                 ::pyo3::impl_::extract_argument::extract_pyclass_ref::<
>                     UUID,
>                 >(
>                     ::pyo3::impl_::pymethods::BoundRef::ref_from_ptr(py, &_slf).0,
>                     &mut holder_0,
>                 )?,
>             ),
>         );
>         result
>     }
2603a2635
> static SAFE_UUID_UNKNOWN: AtomicPtr<ffi::PyObject> = AtomicPtr::new(null_mut());
2650a2683,2694
>     let safe_uuid_unknown = Python::with_gil(|py| {
>         return PyModule::import_bound(py, "uuid")
>             .unwrap()
>             .getattr("SafeUUID")
>             .unwrap()
>             .unbind()
>             .bind(py)
>             .getattr("unknown")
>             .unwrap()
>             .unbind();
>     });
>     SAFE_UUID_UNKNOWN.store(safe_uuid_unknown.into_ptr(), Ordering::Relaxed);
```

**Comment by trim21** [comment]
:eyes:

**Review by aminalaee — APPROVED** [review]
Thanks for the PR 👍

**Pull request merged**